datasync/chaindatafetcher/kafka/consumer.go

```diff
@@ -306,37 +307,99 @@ func (c *Consumer) updateOffset(buffer [][]*Segment, lastMsg *sarama.ConsumerMes
 // loop and exit.
 func (c *Consumer) ConsumeClaim(cgs sarama.ConsumerGroupSession, cgc sarama.ConsumerGroupClaim) error {
-	var buffer [][]*Segment
-	for msg := range cgc.Messages() {
-		if len(buffer) > c.config.MaxMessageNumber {
-			return fmt.Errorf("%v: increasing buffer size may resolve this problem. [max: %v, current: %v]", bufferOverflowErrorMsg, c.config.MaxMessageNumber, len(buffer))
-		}
+	if buffer, err := c.consumeClaim(cgs, cgc); err != nil {
+		return c.handleError(buffer, cgs, err)
+	}
+	return nil
+}
 
-		segment, err := newSegment(msg)
-		if err != nil {
-			return err
-		}
+func (c *Consumer) consumeClaim(cgs sarama.ConsumerGroupSession, cgc sarama.ConsumerGroupClaim) ([][]*Segment, error) {
+	var (
+		buffer      [][]*Segment // TODO-Chaindatafetcher better to introduce segment buffer structure with useful methods
+		expireMsg   *sarama.ConsumerMessage
+		expireTimer = time.NewTimer(c.config.ExpirationTime)
+	)
 
-		// insert a new message segment into the buffer
-		buffer, err = insertSegment(segment, buffer)
-		if err != nil {
-			return err
-		}
+	// make sure that the expireTimer channel is empty
+	if !expireTimer.Stop() {
+		<-expireTimer.C
+	}
 
-		// handle the buffered messages if any message can be reassembled
-		buffer, err = c.handleBufferedMessages(buffer)
-		if err != nil {
-			return err
+	for {
+		select {
+		case <-expireTimer.C:
+			return buffer, errors.New(msgExpiredErrorMsg)
+		case msg, ok := <-cgc.Messages():
+			if !ok {
+				return buffer, nil
+			}
+
+			if len(buffer) > c.config.MaxMessageNumber {
+				return buffer, fmt.Errorf("%v: increasing buffer size may resolve this problem. [max: %v, current: %v]", bufferOverflowErrorMsg, c.config.MaxMessageNumber, len(buffer))
+			}
+
+			segment, err := newSegment(msg)
+			if err != nil {
+				return buffer, err
+			}
+
+			// insert a new message segment into the buffer
+			buffer, err = insertSegment(segment, buffer)
+			if err != nil {
+				return buffer, err
+			}
+
+			// handle the buffered messages if any message can be reassembled
+			buffer, err = c.handleBufferedMessages(buffer)
+			if err != nil {
+				return buffer, err
+			}
+
+			// update expire message and timer if it is handled
+			expireMsg = c.renewExpireMsg(buffer, expireTimer, expireMsg)
+
+			// mark offset of the oldest message to be read
+			if err := c.updateOffset(buffer, msg, cgs); err != nil {
+				return buffer, err
+			}
 		}
+	}
+}
 
-		// mark offset of the oldest message to be read
-		if err := c.updateOffset(buffer, msg, cgs); err != nil {
-			return err
+func (c *Consumer) handleError(buffer [][]*Segment, cgs ConsumerGroupSession, parentErr error) error {
+	if len(buffer) > 0 {
+		oldestMsg := buffer[0][0].orig
+		key := string(oldestMsg.Key)
+
+		if c.config.ErrCallback != nil {
+			if err := c.config.ErrCallback(key); err != nil {
+				return err
+			}
+
+			buffer = buffer[1:]
+			return c.updateOffset(buffer, oldestMsg, cgs)
```

Comment:
The error can occur while consuming a message not placed in buffer[0], but this method always removes buffer[0]. Is that you intended?